File: crates/noidroid-cli/tests/doctor.rs
//! `noidroid doctor`, through the real binary.
//!
//! Automatic capture fails open by construction: every patching mechanism can miss a
//! surface, and a recording that missed one still looks real. The doctor exists to say
//! so *before* a recording is made, which means the claim worth testing is not that it
//! prints a lot of ticks. It is that it never prints one it did not earn.
//!
//! So these tests are all about the same distinction: **we looked and it is not
//! captured** is a different sentence from **we did not look**, and neither of them is
//! `ok`. A doctor that collapses the three is worse than no doctor, because it launders
//! an unexamined surface into a green tick.

use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};

fn repo_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../..")
        .canonicalize()
        .expect("the workspace root is two levels up")
}

fn workdir(tag: &str) -> PathBuf {
    let dir = std::env::temp_dir().join(format!(
        "noidroid-doctor-{tag}-{}-{}",
        std::process::id(),
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos()
    ));
    std::fs::create_dir_all(&dir).unwrap();
    dir
}

fn noidroid(dir: &Path, args: &[&str]) -> (String, bool) {
    let root = repo_root();
    let output = Command::new(env!("CARGO_BIN_EXE_noidroid"))
        .args(args)
        .current_dir(dir)
        .env("PYTHONPATH", root.join("clients/python"))
        .env("NO_COLOR", "1")
        .output()
        .expect("the binary should run");
    let mut text = String::from_utf8_lossy(&output.stdout).to_string();
    text.push_str(&String::from_utf8_lossy(&output.stderr));
    (text, output.status.success())
}

/// The one line of the report about `name`, so a test can assert on what that line
/// does *not* say as well as what it does.
fn line_about<'a>(report: &'a str, name: &str) -> &'a str {
    report
        .lines()
        .map(str::trim_end)
        .find(|line| line.trim_start().starts_with(name))
        .unwrap_or_else(|| panic!("no line about '{name}' in:\n{report}"))
}

fn write(dir: &Path, name: &str, source: &str) -> PathBuf {
    let path = dir.join(name);
    std::fs::write(&path, source).unwrap();
    path
}

fn python_can(expression: &str) -> bool {
    let root = repo_root();
    matches!(
        Command::new("python3")
            .args(["-c", expression])
            .env("PYTHONPATH", root.join("clients/python"))
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status(),
        Ok(status) if status.success()
    )
}

#[test]
fn a_surface_we_did_not_look_at_is_never_reported_as_covered() {
    let dir = workdir("unlooked");

    // No program named, so nothing about the program can have been examined.
    let (report, _) = noidroid(&dir, &["doctor"]);

    for check in ["clock", "subprocess"] {
        let line = line_about(&report, check);
        assert!(
            line.contains("not determined"),
            "'{check}' was not looked at, so it may not be reported as anything else: {line}"
        );
        assert!(
            !line.contains(" ok "),
            "'{check}' must never be ticked when nothing examined it: {line}"
        );
    }
    assert!(
        report.contains("no program given"),
        "the report has to say why it could not look: {report}"
    );
    // The reader must be told that the amber lines are not passes, in the report and
    // not only in the documentation.
    assert!(
        report.contains("not determined is not a pass"),
        "an unexamined surface has to be called what it is: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn a_clean_scan_is_not_reported_as_a_clean_program() {
    let dir = workdir("clean");
    let agent = write(
        &dir,
        "agent.py",
        "import json\n\n\ndef main():\n    return json.dumps({\"ok\": True})\n",
    );

    let (report, _) = noidroid(&dir, &["doctor", "--", "python3", agent.to_str().unwrap()]);

    // One file was parsed and had nothing in it. That is a fact about one file, not a
    // clean bill of health for the program: the scan does not follow imports, so the
    // rest of it was never opened.
    let line = line_about(&report, "clock");
    assert!(
        line.contains("not determined"),
        "a scan that read one file cannot clear a program: {line}"
    );
    assert!(
        report.contains("imports are not followed"),
        "the boundary of the scan has to be stated: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn the_clock_and_randomness_are_named_with_the_line_they_appear_on() {
    let dir = workdir("clock");
    let agent = write(
        &dir,
        "agent.py",
        "import time\nfrom uuid import uuid4\n\n\ndef go(client):\n    \
         return client.messages.create(id=str(uuid4()), at=time.time())\n",
    );

    let (report, _) = noidroid(&dir, &["doctor", "--", "python3", agent.to_str().unwrap()]);

    let line = line_about(&report, "clock");
    assert!(
        line.contains("not captured"),
        "this one we looked at and found, which is not the same as not looking: {line}"
    );
    assert!(
        report.contains("uuid.uuid4") && report.contains("time.time"),
        "the report must name the call, not just its module: {report}"
    );
    assert!(
        report.contains("agent.py:6"),
        "and where it is, so it can be marked volatile= without a search: {report}"
    );
    assert!(
        report.contains("#30"),
        "a known, filed hole is named as one: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn a_subprocess_is_reported_as_a_known_hole_not_as_covered() {
    let dir = workdir("subprocess");
    let agent = write(
        &dir,
        "agent.py",
        "import subprocess\n\n\ndef build():\n    \
         return subprocess.run([\"make\"], check=True)\n",
    );

    let (report, _) = noidroid(&dir, &["doctor", "--", "python3", agent.to_str().unwrap()]);

    let line = line_about(&report, "subprocess");
    assert!(
        line.contains("not captured"),
        "a child process is neither recorded nor fenced, and the doctor says so: {line}"
    );
    assert!(
        report.contains("subprocess.run"),
        "the call has to be named: {report}"
    );
    assert!(
        report.contains("#31"),
        "the filed issue is named rather than the surface implied to be covered: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn an_installed_sdk_names_every_request_surface_and_whether_it_is_hooked() {
    if !python_can("import anthropic") {
        eprintln!("SKIP: doctor SDK test — the Anthropic SDK is not installed");
        return;
    }
    let dir = workdir("sdk");

    let (report, ok) = noidroid(&dir, &["doctor"]);

    // Both surfaces are read back out of the SDK after running the real installer, so
    // this is what is patched rather than what we meant to patch.
    assert!(
        report.contains("anthropic._base_client.SyncAPIClient.request"),
        "the hooked surface is named: {report}"
    );
    assert!(
        report.contains("anthropic._base_client.AsyncAPIClient.request"),
        "and so is the one that is present and not hooked: {report}"
    );
    assert!(
        report.contains("NOT hooked"),
        "an unhooked surface has to be unmissable: {report}"
    );
    assert!(
        report.contains("#33"),
        "the async hole is filed, and is named rather than implied covered: {report}"
    );
    let line = line_about(&report, "anthropic");
    assert!(
        line.contains("blocked"),
        "an installed SDK with an unpatched surface is a hard fail: {line}"
    );
    assert!(
        !ok,
        "and a blocked check must not exit zero, or CI would not notice it: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn the_fence_is_reported_only_after_it_actually_refuses_a_connection() {
    let dir = workdir("fence");

    let (report, _) = noidroid(&dir, &["doctor"]);

    let line = line_about(&report, "egress");
    assert!(
        line.contains(" ok"),
        "the fence really did stop a connect, so it may be ticked: {line}"
    );
    assert!(
        report.contains("203.0.113.1:80"),
        "the tick is earned by a named, refused connection, not by an import: {report}"
    );
    assert!(
        report.contains("blind to"),
        "what the fence cannot see is printed next to what it can: {report}"
    );
    assert!(
        report.contains("#31"),
        "including the filed one — a fence blind to subprocesses is not a fence: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn a_client_version_that_cannot_be_read_is_not_reported_as_matching() {
    if python_can("from importlib.metadata import version; version('noidroid')") {
        eprintln!("SKIP: doctor version test — the client is installed, so its version reads");
        return;
    }
    let dir = workdir("version");

    let (report, _) = noidroid(&dir, &["doctor"]);

    let line = line_about(&report, "version");
    assert!(
        line.contains("not determined"),
        "a version nothing could read is not a version that matches: {line}"
    );
    assert!(
        !line.contains("matches"),
        "and must not be described as one: {line}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn the_report_never_prints_a_score_a_percentage_or_a_grade() {
    let dir = workdir("score");

    // No program named, so no filesystem paths land in the output to confuse this.
    let (report, _) = noidroid(&dir, &["doctor"]);

    assert!(
        !report.contains('%'),
        "a percentage of a program's surfaces is a number nobody measured: {report}"
    );
    for word in ["score", "grade", "readiness", "healthy", "confidence"] {
        assert!(
            !report.to_lowercase().contains(word),
            "'{word}' turns a list of findings into fidelity theatre: {report}"
        );
    }
    // Filesystem paths carry slashes of their own, and none of them is a grade.
    let root = repo_root().display().to_string();
    let ratio = report
        .lines()
        .filter(|line| !line.contains(&root))
        .flat_map(|line| {
            line.chars()
                .collect::<Vec<_>>()
                .windows(3)
                .map(|w| w[0].is_ascii_digit() && w[1] == '/' && w[2].is_ascii_digit())
                .collect::<Vec<_>>()
        })
        .any(|hit| hit);
    assert!(
        !ratio,
        "'4 of 6 checks passed' is a grade with extra steps: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

#[test]
fn a_preflight_does_not_create_a_repository_just_by_asking() {
    let dir = workdir("norepo");

    let (report, _) = noidroid(&dir, &["doctor"]);

    assert!(
        !dir.join(".noidroid").exists(),
        "asking what would be recorded must record nothing, not even a store: {report}"
    );

    let _ = std::fs::remove_dir_all(&dir);
}

File: clients/python/noidroid/doctor.py
"""What a recording would and would not cover, probed rather than assumed.

`noidroid doctor` needs answers about a Python process it is not inside: which SDKs
are importable, which of their request surfaces automatic capture actually patched,
whether the egress fence really refuses a connection. Deciding any of that from Rust
would produce a report about *intent* — the list of things we meant to hook — and the
whole point of a preflight here is that automatic capture fails open, so intent is
exactly the thing that must not be trusted.

So this module runs the real `noidroid.auto.install` and `noidroid.fence.install` in a
throwaway process, then reads back what is actually patched and whether a connect is
actually refused. It also parses the program you are about to record, looking for the
clock, randomness and subprocesses — the three holes that make a replay diverge or,
worse, quietly miss what a child process did.

It reports facts and no verdicts. Whether a fact is a pass, a known hole, or something
nobody looked at is decided in one place, by the doctor, because that judgement is the
part that has to stay honest.

Output is one JSON object on stdout, behind a sentinel so that an SDK which prints on
import cannot corrupt it. Run it as `python3 -m noidroid.doctor [command...]`.
"""

from __future__ import annotations

import ast
import importlib
import json
import os
import socket
import sys
from typing import Any, Optional

__all__ = ["main", "SENTINEL"]

#: The report is the last line beginning with this. Anything an SDK printed on import
#: lands above it and is ignored rather than parsed as half a report.
SENTINEL = "__noidroid_doctor__ "

#: An address in TEST-NET-3 (RFC 5737), which is reserved for documentation and is not
#: routed. The fence is supposed to refuse it before a packet exists; the short timeout
#: is what keeps a *broken* fence from hanging this probe for a minute.
_FENCE_TARGET = ("203.0.113.1", 80)
_FENCE_TIMEOUT = 0.2

# --------------------------------------------------------------- non-determinism

#: Calls whose value differs between two runs of the same program. Named exactly,
#: because `x.now()` on something that is not a datetime is not a clock read.
_CLOCK = {
    "time.time",
    "time.time_ns",
    "time.monotonic",
    "time.monotonic_ns",
    "time.perf_counter",
    "time.perf_counter_ns",
    "time.localtime",
    "time.gmtime",
    "time.asctime",
    "time.ctime",
    "time.strftime",
    "datetime.datetime.now",
    "datetime.datetime.utcnow",
    "datetime.datetime.today",
    "datetime.datetime.fromtimestamp",
    "datetime.date.today",
}
_RANDOM = {"uuid.uuid1", "uuid.uuid4", "os.urandom", "os.getpid"}
_RANDOM_PREFIX = ("random.", "secrets.")
_SUBPROCESS = {"os.system", "os.popen", "multiprocessing.Process", "pty.spawn"}
_SUBPROCESS_PREFIX = ("subprocess.", "asyncio.create_subprocess", "os.spawn", "os.exec")
#: Importing one of these is itself the signal (#31): a child process is neither
#: recorded nor fenced, so it does not have to be called to be worth reporting.
_SUBPROCESS_MODULES = {"subprocess", "pty"}


def _classify(name: str) -> Optional[str]:
    if name in _CLOCK:
        return "clock"
    if name in _RANDOM or name.startswith(_RANDOM_PREFIX):
        return "randomness"
    if name in _SUBPROCESS or name.startswith(_SUBPROCESS_PREFIX):
        return "subprocess"
    return None


class _Sites(ast.NodeVisitor):
    """Every site in one file where a value we do not capture enters the program.

    Import bindings are tracked so that `from uuid import uuid4` and `import uuid`
    resolve to the same name. Without that, a scan either misses the first form or
    flags every bare `uuid4` a program happens to define, and a preflight full of
    false alarms gets ignored, which is the same outcome as not having one.
    """

    def __init__(self, path: str) -> None:
        self.path = path
        self.alias: dict[str, str] = {}
        self.findings: list[dict[str, Any]] = []

    def _record(self, line: int, name: str, kind: str) -> None:
        self.findings.append(
            {"file": self.path, "line": line, "name": name, "kind": kind}
        )

    def visit_Import(self, node: ast.Import) -> None:
        for entry in node.names:
            root = entry.name.split(".")[0]
            self.alias[entry.asname or root] = entry.name if entry.asname else root
            if root in _SUBPROCESS_MODULES:
                self._record(node.lineno, f"import {entry.name}", "subprocess")
        self.generic_visit(node)

    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
        module = node.module or ""
        for entry in node.names:
            if entry.name == "*":
                continue
            self.alias[entry.asname or entry.name] = (
                f"{module}.{entry.name}" if module else entry.name
            )
        if module.split(".")[0] in _SUBPROCESS_MODULES:
            self._record(node.lineno, f"from {module} import …", "subprocess")
        self.generic_visit(node)

    def visit_Call(self, node: ast.Call) -> None:
        name = self._resolve(node.func)
        if name is not None:
            kind = _classify(name)
            if kind is not None:
                self._record(node.lineno, name, kind)
        self.generic_visit(node)

    def _resolve(self, node: ast.expr) -> Optional[str]:
        parts: list[str] = []
        while isinstance(node, ast.Attribute):
            parts.append(node.attr)
            node = node.value
        if not isinstance(node, ast.Name):
            return None
        parts.append(node.id)
        parts.reverse()
        head = self.alias.get(parts[0], parts[0])
        return ".".join([head] + parts[1:])


def scan(arguments: list) -> dict:
    """Parse whichever arguments name a readable Python file.

    Imports are deliberately not followed. Doing so means resolving a package layout
    we do not own and reporting on files the user did not point at; not doing so means
    a clean result covers one file and no more. The second is a smaller lie, and the
    doctor states the boundary rather than hiding it.

    Used two ways: as one section of the full probe below, and — behind `--scan`, with
    none of `_providers()` or `_fence()` run — as the whole report a recording asks for
    of itself once it finishes (#71). The parse is the same either way, on purpose:
    there is exactly one place that decides what counts as a clock, a randomness read
    or a subprocess launch, and a recording seeing something doctor would not is a bug
    worth finding, not a second opinion.
    """
    scanned: list[str] = []
    unreadable: list[dict[str, str]] = []
    findings: list[dict[str, Any]] = []
    for argument in arguments:
        if not argument.endswith(".py") or not os.path.isfile(argument):
            continue
        try:
            with open(argument, "r", encoding="utf-8") as handle:
                source = handle.read()
            tree = ast.parse(source, filename=argument)
        except (OSError, SyntaxError, ValueError) as exc:
            unreadable.append({"file": argument, "error": f"{type(exc).__name__}: {exc}"})
            continue
        sites = _Sites(argument)
        sites.visit(tree)
        scanned.append(argument)
        findings.extend(sites.findings)
    return {"scanned": scanned, "unreadable": unreadable, "findings": findings}


# ------------------------------------------------------------------- the client


def _client() -> dict:
    import noidroid

    info: dict[str, Any] = {"path": getattr(noidroid, "__file__", None), "version": None}
    try:
        from importlib.metadata import version

        info["version"] = version("noidroid")
    except Exception:
        # Importable from a source tree on PYTHONPATH rather than installed. The
        # version is then genuinely unknown, and the doctor says so rather than
        # assuming it matches.
        pass
    return info


# ---------------------------------------------------------------- the surfaces


def _providers() -> list:
    """Run the real installer, then read back which surfaces carry the patch.

    Enumerating `*APIClient` classes rather than trusting `auto`'s own list is the
    point: a surface this build has never heard of still shows up here, unhooked,
    instead of being invisible to both the patcher and the report.
    """
    from noidroid import auto

    out = []
    for provider in auto.PROVIDERS:
        info: dict[str, Any] = {
            "name": provider,
            "installed": False,
            "version": None,
            "surfaces": [],
            "error": None,
        }
        try:
            module = importlib.import_module(provider)
        except ImportError:
            out.append(info)
            continue
        info["installed"] = True
        info["version"] = getattr(module, "__version__", None)
        try:
            auto.install((provider,))
        except Exception as exc:  # noqa: BLE001
            info["error"] = f"{type(exc).__name__}: {exc}"
        try:
            base = importlib.import_module(f"{provider}._base_client")
        except ImportError as exc:
            info["error"] = info["error"] or f"{type(exc).__name__}: {exc}"
            out.append(info)
            continue
        for attribute in sorted(dir(base)):
            if not attribute.endswith("APIClient"):
                continue
            client_cls = getattr(base, attribute)
            if not isinstance(client_cls, type):
                continue
            request = getattr(client_cls, "request", None)
            if request is None:
                continue
            info["surfaces"].append(
                {
                    "name": f"{provider}._base_client.{attribute}.request",
                    "hooked": bool(getattr(request, "__noidroid_wrapped__", False)),
                }
            )
        out.append(info)
    return out


# -------------------------------------------------------------------- the fence


def _fence() -> dict:
    """Install the fence for real and check it refuses a connection.

    Reporting that the module imported would be reporting on intent. The fence is
    only worth anything if a non-local connect raises, so that is what is tested —
    and the guard raises before the real `connect`, so nothing leaves the machine.
    """
    result: dict[str, Any] = {
        "installed": False,
        "refused": None,
        "target": f"{_FENCE_TARGET[0]}:{_FENCE_TARGET[1]}",
        "error": None,
    }
    try:
        from noidroid import fence

        fence.install(strict=True)
        result["installed"] = bool(
            getattr(socket.socket.connect, "__noidroid_fenced__", False)
        )
        probe = socket.socket()
        probe.settimeout(_FENCE_TIMEOUT)
        try:
            probe.connect(_FENCE_TARGET)
            result["refused"] = False
        except fence.Escaped:
            result["refused"] = True
        except OSError as exc:
            result["refused"] = False
            result["error"] = f"{type(exc).__name__}: {exc}"
        finally:
            probe.close()
    except Exception as exc:  # noqa: BLE001
        result["error"] = f"{type(exc).__name__}: {exc}"
    return result


def main(argv: Optional[list] = None) -> int:
    arguments = list(sys.argv[1:] if argv is None else argv)
    if arguments[:1] == ["--scan"]:
        # The lean path: parse the files named and say what is in them, and nothing
        # about the SDKs installed or the fence in *this* throwaway process. A
        # recording asks this after every run (#71); running `_providers()` and
        # `_fence()` on every recording would import and patch SDKs, and install a
        # socket fence, for a question that is purely about source text.
        report = scan(arguments[1:])
        sys.stdout.write(SENTINEL + json.dumps(report) + "\n")
        sys.stdout.flush()
        return 0
    report = {
        "client": _client(),
        "providers": _providers(),
        "scan": scan(arguments),
        # Last, because it patches this process's sockets and everything above it
        # may legitimately want one.
        "fence": _fence(),
    }
    sys.stdout.write(SENTINEL + json.dumps(report) + "\n")
    sys.stdout.flush()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
